openedx/features/discounts/docs/decisions/0001-determining-discount-location.rst

```diff
@@ -0,0 +1,50 @@
+Course + User Discounts
+-----------------------------
+
+Status
+======
+
+Accepted
+
+Context
+=======
+
+We are implementing platform-wide discounts that are determined using a combination of course and user. The
+data used in this determination lives in the lms. The discount will be messaged in the lms, and then autoapplied
+on the basket page of ecommerce.
+
+Decisions
+=========
+
+#. **The location of the calculation of discount applicability/eligibility**
+
+   a. We want *one* service to contain the logic for whether a discount can be applied/can be communicated to the 
+      user. This way, upkeep is much simpler: if/when the logic changes, it only needs to change in one place.
+      Additionally, if we duplicated the logic, we would either need to duplicate the data, ask for the data, or
+      have slightly different logic (which would present cases in which the outcome is different depending on the
+      service). We are unhappy with all of those implementations.
+
+   b. We will calculate applicability/eligibility in the lms. All of the data that we need currently resides in 
+      the LMS. In order to get the information to make the determination in ecommerce, we would need to ask LMS
+      for the data. If the ultimate destination of that data was the LMS (to decide whether to message the 
+      discount), we would have excess communication and more failure points.
+
+#. **Communication of discount applicability**
+
+   a. Within the LMS, when we need to get the applicability of a discount for a user and course, we can just
+      import the function and call it.
+
+   b. Outside of the LMS, we need a way to get this information in a validated fashion. We have added a rest 
+      endpoint for that communication. Because we want to be able to use the information both on the front-end
+      and back-end without risk of the user creating their own discounts, we are sending it as a jwt, which is
+      signed. See C for more information on how specifically it will be used on the ecommerce front-end.
+
+   c. We do not want to make a synchronous call from ecommerce to the LMS every time the basket page is loaded.
```

Comment:
could we clarify why we don't want to do this? it sounds like we would be making an async call and reloading the page for most users, which would of course perform much worse for the user than a synchronous call

Reply:
Hmm...Everyone I talked to at edX about this thought of it as a given. Sure. I can flesh it out.